README.md
Uri -- an RFC3986 URI/URL parsing library
-----------------------------------------

This is an OCaml implementation of the [RFC3986](http://tools.ietf.org/html/rfc3986) specification 
for parsing URI or URLs.

## Installation

### Via OPAM

The [OPAM](https://opam.ocaml.org) package manager can be used to install this library from source.

    opam install uri

### Locally

You can build the source code locally via the [dune](https://github.com/ocaml/dune)
build system.

    opam install uri --deps-only
    eval `opam config env`
    dune build
    dune runtest

will install the dependencies via OPAM, build the library and then run the tests in the [lib_test/](lib_test/) directory.

## Usage

Once installed, there are three ocamlfind packages available for your use:

- `uri` - the base `Uri` module
- `uri-re` - the _legacy_ implementation
  At the beginning, `uri` used `re` to parse a `string`. Since 4.0.0,
  we use `angstrom` - if something breaks with `uri.4.0.0`, you should compare
  with `uri-re` and submit an issue. `uri-re` is deprecated and it will be
  removed on the next release (see #150)
- `uri.top` - the toplevel printers for use with [utop](https://github.com/diml/utop)
- `uri-sexp` - provides converters to and from s-expressions (via a `Uri_sexp.t` type alias)
- `uri.services` - the `Uri_services` module that provides the equivalent of *[services(5)](http://man7.org/linux/man-pages/man5/services.5.html)*
- `uri.services_full` - the `Uri_services_full` module that provides a complete copy of the `/etc/services` file. This is quite large and normally not needed.

## Contact

- Issues: <https://github.com/mirage/ocaml-uri/issues>
- E-mail: <mirageos-devel@lists.xenproject.org>
- API Documentation: <https://ocaml.org/p/uri/latest/doc/index.html>

[![Build Status](https://travis-ci.org/mirage/ocaml-uri.png)](https://travis-ci.org/mirage/ocaml-uri)


CHANGES.md
v4.3.0 2023-04-19
-----------------

* Add `Uri.Absolute_http`, an RFC9110-compliance specialization of a
  `Uri.t`. (#164 #162 @torinnd).
* Add a `uri-bench` package for the benchmarking dependencies in this
  repository (#166 @tmcgilchrist).

v4.2.0 2021-04-26
-----------------

* Do not mutate the base encoder when using custom percent encoders.
  This was a bug introduced in #147. (#156 @aantron)
* Disable Travis CI tests and switch Win/Mac tests to GitHub Actions and
  Linux ones to ocaml-ci (@avsm).

v4.1.0 2021-01-04
-----------------

* `uri-re` is deprecated, it is a legacy implementation that is now outdated. 
  `uri` should be used instead (@dinosaure, #152).
* Fix build system for cross-compilation (@TheLortex, #151).

v4.0.0 2020-01-01
-----------------

* sexp: use the sexplib v0.13 ppx directives (@avsm, #143).
* rework the URI parser with `angstrom` (@anmonteiro, review @avsm & @dinosaure, #142).
* add simple fuzzer tests between `angstrom` parser and _legacy_ parser (with `re.posix`, #142)
* add support of modifying pct encoding (with a custom one) (@orbitz, review @anmonteiro, @tmcgilchrist, @avsm & @dinosaure, #147)
* allow the selection of generic set of safe characters (with `Generic`) (@madroach, review @dinosaure, #141)

v3.1.0 2019-11-23
-----------------

* Add `Uri_sexp.equal` (#139 @vbmithr)
* Update opam files to remove dune as a build-only dep (@craigfe #140)

v3.0.0 2019-07-06
-----------------

* Complete the migration of making sexp an optional dependency that was
  started in 2.0.0. We now remove the `uri.sexp` ocamlfind package and
  have `uri` and `uri-sexp` for both the ocamlfind and opam packages.
  Code that was formerly using `uri.sexp` in its build will now need to
  move to `uri-sexp` instead (#134 @Julow @dinosaure).

* Remove the deprecated `Uri_re` module. All code should be using the
  `Uri.Re` module instead (@avsm @Julow).

* Remove the `uri.top` library, since we install the toplevel printer
  automatically since 2.2.0 via an attribute.

v2.2.1 2019-06-02
-----------------

* Fix deprecation warnings in Re 1.9.0 (#137 @avsm).

v2.2.0 2019-01-31
-----------------

* Add `Uri.pp` as an alias to `Uri.pp_hum`, as the `pp` form
  is more commonly used. (#133 @avsm)
* Add an `[@@ocaml.toplevel_printer]` attribute to Uri.pp
  so that it will be automatically loaded on modern Utop versions. (#133 @avsm)
* Upgrade last remaining `jbuild` file to `dune` (#133 @avsm)
* OCamldoc improvements in section headers (@avsm)

v2.1.0 2018-12-12
-----------------

* Expose a `compare` function in `Uri_sexp` so that it will work
  with `deriving compare,sexp`.
* Upgrade the opam metadata to the 2.0 format.
* Update Travis to test OCaml 4.03->4.07.
* Minimum OCaml version is now 4.04.0+ due to sexplib0 dependency.

v2.0.0 2018-10-15
-----------------

* Create a new subpackage `uri.sexp` for the sexpression converters, so
  that the base Uri package is more dependency free (based on #121 by @Julow).
  To convert old code, simply use `Uri_sexp.t` instead of `Uri.t` in a record
  type for which you are using `[@@deriving sexp]`.  This is a type alias to
  `Uri.t` but also has the right sexp-conversion functions in scope.
* Deprecate `Uri_re` in favour of `Uri.Re`.  The unwrapped `Uri_re` will
  be removed in a future release of this library.
* Switch to using sexplib0 instead of sexplib for easier dependencies
  (based on #123 by @mseri)
* Port build to Dune from jbuilder.
* Add benchmarks using `core_bench` (#125 via @drup)

v1.9.7 2018-07-01
-----------------

* Use latest interfaces in Re >=1.7.2 (#122)
* When resolving URI inherit userinfo from the base URI (#116)

v1.9.6 2018-01-11
-----------------

* Change code generation strategy to avoid big switches in
  the services file; improves build time by 10x (#114 by @gasche).
* Remove deprecated function use (`String.lowercase`)
* Add development Makefile with more targets.

v1.9.5 2017-11-05
-----------------

* Fix build with OCaml 4.06 (and -safe-string) (#108 @hcarty)
* Set (wrapped false) in jbuilder (#105 @avsm)
* Add OCaml 4.06 to the travis CI matrix (#109 @djs55)

v1.9.4 2017-05-30
-----------------

* Port build system to jbuilder (#100 @vbmithr @rgrinberg @avsm @dsheets).
  There should be no observable changes, except that `Uri_services` is now
  in a separate subdirectory. This means that packages that implicitly
  depended on the module without including the ocamlfind `uri.services`
  package may now fail. Just adding the ocamlfind dependency will fix it,
  and is backwards compatible with older Uri releases.
* Restrict build to OCaml 4.03.0+ (was formerly OCaml 4.02.0+).
* Add Appveyor tests for Windows compilation.

v1.9.3 2017-03-06
-----------------

* Port build system to topkg (#95 by @fgimenez)
* Add a tighter opam constraint on `ppx_sexp` (#94)
* Explicitly depend on `ppx_deriving` for improving future compatibility
  with Jane Street upstream (#98).
* Update Travis CI to include OCaml 4.04 and 4.03 in the matrix.

v1.9.2 2016-02-12
-----------------

* Remove sexplib.syntax, `type_conv` deps and camlp4 transitive dependency
* Add `ppx_sexp_conv` dependency
* Require OCaml 4.02.3+

v1.9.1 2015-06-26
-----------------

* Fix `with_password None` when no userinfo present (#78 from Hezekiah M. Carty)

v1.9.0 2015-05-15
-----------------

* Colon (":") is no longer percent-encoded in path segments
* URNs are now supported (#67)
* Relative paths with colons in first segment have "./" prepended in to_string
* Add Uri.empty, the zero length URI reference
* `Uri_services` now includes service aliases (e.g. www, www-http, http)
* `Uri_services` now includes chargen and git
* Add `Uri.canonicalize` for scheme-specific normalization (#70)
* Add `Uri.verbatim_query` to extract literal query string (#57)
* Add `Uri.equal`
* Add `Uri.user` and `Uri.password` accessors for subcomponents of userinfo (#62)
* Add `Uri.with_password` functional setter for password subcomponent of userinfo

lib/uri_services.ml
(* Autogenerated by gen_services.ml, do not edit directly *)
let tcp_port_of_service = function
  |"9pfs" -> [564]
  |"acap" -> [674]
  |"auth" -> [113]
  |"bgp" -> [179]
  |"bootpc" -> [68]
  |"bootps" -> [67]
  |"chargen" -> [19]
  |"daytime" -> [13]
  |"dctp" -> [675]
  |"dhcpv6-client" -> [546]
  |"dhcpv6-server" -> [547]
  |"discard" -> [9]
  |"domain" -> [53]
  |"echo" -> [7]
  |"efs" -> [520]
  |"finger" -> [79]
  |"ftp" -> [21]
  |"ftp-data" -> [20]
  |"ftps" -> [990]
  |"ftps-data" -> [989]
  |"genrad-mux" -> [176]
  |"git" -> [9418]
  |"gopher" -> [70]
  |"http" -> [80]
  |"https" -> [443]
  |"ident" -> [113]
  |"imap" -> [143]
  |"imap4-ssl" -> [585]
  |"imaps" -> [993]
  |"ipx" -> [213]
  |"irc" -> [194]
  |"irc-serv" -> [529]
  |"ircs" -> [994]
  |"kerberos" -> [88]
  |"kerberos-adm" -> [749]
  |"keyserver" -> [584]
  |"klogin" -> [543]
  |"kpasswd" -> [464]
  |"kshell" -> [544]
  |"ldap" -> [389]
  |"msp" -> [18]
  |"netbios-dgm" -> [138]
  |"netbios-ns" -> [137]
  |"netbios-ssn" -> [139]
  |"nmap" -> [689]
  |"nntp" -> [119]
  |"ntalk" -> [518]
  |"ntp" -> [123]
  |"pirp" -> [553]
  |"pop3" -> [110]
  |"pop3s" -> [995]
  |"printer" -> [515]
  |"qmtp" -> [209]
  |"rsync" -> [873]
  |"rtsp" -> [554]
  |"silc" -> [706]
  |"smtp" -> [25]
  |"ssh" -> [22]
  |"sunrpc" -> [111]
  |"syslog-conn" -> [601]
  |"systat" -> [11]
  |"talk" -> [517]
  |"telnet" -> [23]
  |"telnets" -> [992]
  |"tftp" -> [69]
  |"time" -> [37]
  |"utime" -> [519]
  |"www" -> [80]
  |"www-http" -> [80]
  |"xdmcp" -> [177]
  |_ -> []

let udp_port_of_service = function
  |"9pfs" -> [564]
  |"acap" -> [674]
  |"auth" -> [113]
  |"bgp" -> [179]
  |"bootpc" -> [68]
  |"bootps" -> [67]
  |"daytime" -> [13]
  |"dctp" -> [675]
  |"dhcpv6-client" -> [546]
  |"dhcpv6-server" -> [547]
  |"discard" -> [9]
  |"domain" -> [53]
  |"echo" -> [7]
  |"finger" -> [79]
  |"ftp" -> [21]
  |"ftp-data" -> [20]
  |"ftps" -> [990]
  |"ftps-data" -> [989]
  |"git" -> [9418]
  |"gopher" -> [70]
  |"http" -> [80]
  |"https" -> [443]
  |"imap" -> [143]
  |"imap4-ssl" -> [585]
  |"imaps" -> [993]
  |"ipx" -> [213]
  |"irc" -> [194]
  |"irc-serv" -> [529]
  |"ircs" -> [994]
  |"kerberos" -> [88]
  |"kerberos-adm" -> [749]
  |"keyserver" -> [584]
  |"klogin" -> [543]
  |"kpasswd" -> [464]
  |"kshell" -> [544]
  |"ldap" -> [389]
  |"msp" -> [18]
  |"netbios-dgm" -> [138]
  |"netbios-ns" -> [137]
  |"netbios-ssn" -> [139]
  |"nmap" -> [689]
  |"nntp" -> [119]
  |"ntalk" -> [518]
  |"ntp" -> [123]
  |"pirp" -> [553]
  |"pop3" -> [110]
  |"pop3s" -> [995]
  |"printer" -> [515]
  |"qmtp" -> [209]
  |"rsync" -> [873]
  |"rtsp" -> [554]
  |"silc" -> [706]
  |"smtp" -> [25]
  |"ssh" -> [22]
  |"sunrpc" -> [111]
  |"syslog" -> [514]
  |"syslog-conn" -> [601]
  |"systat" -> [11]
  |"talk" -> [517]
  |"telnet" -> [23]
  |"telnets" -> [992]
  |"tftp" -> [69]
  |"time" -> [37]
  |"utime" -> [519]
  |"www" -> [80]
  |"www-http" -> [80]
  |"xdmcp" -> [177]
  |_ -> []

let service_of_tcp_port = function
  |7 -> ["echo"]
  |9 -> ["discard"]
  |11 -> ["systat"]
  |13 -> ["daytime"]
  |18 -> ["msp"]
  |19 -> ["chargen"]
  |20 -> ["ftp-data"]
  |21 -> ["ftp"]
  |22 -> ["ssh"]
  |23 -> ["telnet"]
  |25 -> ["smtp"]
  |37 -> ["time"]
  |53 -> ["domain"]
  |67 -> ["bootps"]
  |68 -> ["bootpc"]
  |69 -> ["tftp"]
  |70 -> ["gopher"]
  |79 -> ["finger"]
  |80 -> ["www-http";"www";"http"]
  |88 -> ["kerberos"]
  |110 -> ["pop3"]
  |111 -> ["sunrpc"]
  |113 -> ["ident";"auth"]
  |119 -> ["nntp"]
  |123 -> ["ntp"]
  |137 -> ["netbios-ns"]
  |138 -> ["netbios-dgm"]
  |139 -> ["netbios-ssn"]
  |143 -> ["imap"]
  |176 -> ["genrad-mux"]
  |177 -> ["xdmcp"]
  |179 -> ["bgp"]
  |194 -> ["irc"]
  |209 -> ["qmtp"]
  |213 -> ["ipx"]
  |389 -> ["ldap"]
  |443 -> ["https"]
  |464 -> ["kpasswd"]
  |515 -> ["printer"]
  |517 -> ["talk"]
  |518 -> ["ntalk"]
  |519 -> ["utime"]
  |520 -> ["efs"]
  |529 -> ["irc-serv"]
  |543 -> ["klogin"]
  |544 -> ["kshell"]
  |546 -> ["dhcpv6-client"]
  |547 -> ["dhcpv6-server"]
  |553 -> ["pirp"]
  |554 -> ["rtsp"]
  |564 -> ["9pfs"]
  |584 -> ["keyserver"]
  |585 -> ["imap4-ssl"]
  |601 -> ["syslog-conn"]
  |674 -> ["acap"]
  |675 -> ["dctp"]
  |689 -> ["nmap"]
  |706 -> ["silc"]
  |749 -> ["kerberos-adm"]
  |873 -> ["rsync"]
  |989 -> ["ftps-data"]
  |990 -> ["ftps"]
  |992 -> ["telnets"]
  |993 -> ["imaps"]
  |994 -> ["ircs"]
  |995 -> ["pop3s"]
  |9418 -> ["git"]
  |_ -> []

let service_of_udp_port = function
  |7 -> ["echo"]
  |9 -> ["discard"]
  |11 -> ["systat"]
  |13 -> ["daytime"]
  |18 -> ["msp"]
  |20 -> ["ftp-data"]
  |21 -> ["ftp"]
  |22 -> ["ssh"]
  |23 -> ["telnet"]
  |25 -> ["smtp"]
  |37 -> ["time"]
  |53 -> ["domain"]
  |67 -> ["bootps"]
  |68 -> ["bootpc"]
  |69 -> ["tftp"]
  |70 -> ["gopher"]
  |79 -> ["finger"]
  |80 -> ["www-http";"www";"http"]
  |88 -> ["kerberos"]
  |110 -> ["pop3"]
  |111 -> ["sunrpc"]
  |113 -> ["auth"]
  |119 -> ["nntp"]
  |123 -> ["ntp"]
  |137 -> ["netbios-ns"]
  |138 -> ["netbios-dgm"]
  |139 -> ["netbios-ssn"]
  |143 -> ["imap"]
  |177 -> ["xdmcp"]
  |179 -> ["bgp"]
  |194 -> ["irc"]
  |209 -> ["qmtp"]
  |213 -> ["ipx"]
  |389 -> ["ldap"]
  |443 -> ["https"]
  |464 -> ["kpasswd"]
  |514 -> ["syslog"]
  |515 -> ["printer"]
  |517 -> ["talk"]
  |518 -> ["ntalk"]
  |519 -> ["utime"]
  |529 -> ["irc-serv"]
  |543 -> ["klogin"]
  |544 -> ["kshell"]
  |546 -> ["dhcpv6-client"]
  |547 -> ["dhcpv6-server"]
  |553 -> ["pirp"]
  |554 -> ["rtsp"]
  |564 -> ["9pfs"]
  |584 -> ["keyserver"]
  |585 -> ["imap4-ssl"]
  |601 -> ["syslog-conn"]
  |674 -> ["acap"]
  |675 -> ["dctp"]
  |689 -> ["nmap"]
  |706 -> ["silc"]
  |749 -> ["kerberos-adm"]
  |873 -> ["rsync"]
  |989 -> ["ftps-data"]
  |990 -> ["ftps"]
  |992 -> ["telnets"]
  |993 -> ["imaps"]
  |994 -> ["ircs"]
  |995 -> ["pop3s"]
  |9418 -> ["git"]
  |_ -> []

let known_tcp_services =
  [ "ftp";"https";"kpasswd";"imap4-ssl";"echo";"netbios-ssn";"www";"pop3";"bgp";"keyserver";"tftp";"rsync";"qmtp";"talk";"kerberos-adm";"ssh";"time";"syslog-conn";"telnets";"printer";"systat";"git";"www-http";"irc-serv";"ldap";"9pfs";"sunrpc";"xdmcp";"kerberos";"genrad-mux";"ident";"pop3s";"pirp";"nntp";"imap";"domain";"netbios-ns";"dctp";"daytime";"msp";"efs";"ftp-data";"http";"kshell";"rtsp";"ftps-data";"chargen";"smtp";"irc";"bootps";"acap";"dhcpv6-client";"finger";"utime";"imaps";"netbios-dgm";"ntp";"nmap";"telnet";"klogin";"silc";"dhcpv6-server";"ircs";"gopher";"auth";"ntalk";"discard";"bootpc";"ipx";"ftps" ]

let known_udp_services =
  [ "ftp";"https";"kpasswd";"imap4-ssl";"echo";"netbios-ssn";"www";"pop3";"bgp";"keyserver";"tftp";"rsync";"qmtp";"talk";"kerberos-adm";"ssh";"time";"syslog-conn";"telnets";"printer";"systat";"git";"www-http";"irc-serv";"ldap";"9pfs";"sunrpc";"xdmcp";"kerberos";"pop3s";"pirp";"nntp";"imap";"domain";"netbios-ns";"syslog";"dctp";"daytime";"msp";"ftp-data";"http";"kshell";"rtsp";"ftps-data";"smtp";"irc";"bootps";"acap";"dhcpv6-client";"finger";"utime";"imaps";"netbios-dgm";"ntp";"nmap";"telnet";"klogin";"silc";"dhcpv6-server";"ircs";"gopher";"auth";"ntalk";"discard";"bootpc";"ipx";"ftps" ]

let known_services = [
  ("tcp", known_tcp_services);
  ("udp", known_udp_services) ]

let port_of_uri ?default lookupfn uri =
  match Uri.port uri with
  |Some port as x -> x
  |None -> begin
    match Uri.scheme uri, default with
    |None, None -> None
    |None, Some scheme
    |Some scheme, _ -> begin
      match lookupfn scheme with
      |[] -> None
      |hd::_ -> Some hd
     end
  end

let tcp_port_of_uri ?default uri =
  port_of_uri ?default tcp_port_of_service uri

let udp_port_of_uri ?default uri =
  port_of_uri ?default udp_port_of_service uri
